libraries/AP_Quicktune/AP_Quicktune.cpp

```diff
@@ -0,0 +1,660 @@
+#include "AP_Quicktune.h"
+
+#if QUICKTUNE_ENABLED
+
+AP_Quicktune *AP_Quicktune::singleton;
+
+const AP_Param::GroupInfo AP_Quicktune::var_info[] = {
+    // @Param: QUIK_ENABLE
+    // @DisplayName: Enable Quicktune.
+    // @Description:
+    // @User: Standard
+    AP_GROUPINFO_FLAGS("ENABLE", 1, AP_Quicktune, enable, 0, AP_PARAM_FLAG_ENABLE),
+
+    // @Param: QUIK_AXES
+    // @DisplayName: Bitmask of axes to tune.
+    // @Description:
+    // @User: Standard
+    AP_GROUPINFO("AXES", 2, AP_Quicktune, axes_enabled, 7),
+
+    // @Param: QUIK_DOUBLE_TIME
+    // @DisplayName: Double time.
+    // @Description: Time taken to double the gain.
+    // @User: Standard
+    AP_GROUPINFO("DOUBLE_TIME", 3, AP_Quicktune, double_time, 10),
+
+    // @Param: QUIK_GAIN_MARGIN
+    // @DisplayName: Gain margin.
+    // @Description:
+    // @User: Standard
+    AP_GROUPINFO("GAIN_MARGIN", 4, AP_Quicktune, gain_margin, 60),
+
+    // @Param: QUIK_OSC_SMAX
+    // @DisplayName: Slewrate for oscillation detection.
+    // @Description:
+    // @User: Standard
+    AP_GROUPINFO("OSC_SMAX", 5, AP_Quicktune, osc_smax, 5),
+
+    // @Param: QUIK_YAW_P_MAX
+    // @DisplayName: Yaw P max.
+    // @Description: Maximum Yaw P gain to tune to.
+    // @User: Standard
+    AP_GROUPINFO("YAW_P_MAX", 6, AP_Quicktune, yaw_p_max, 0.5),
+
+    // @Param: QUIK_YAW_D_MAX
+    // @DisplayName: Yaw D max.
+    // @Description: Maximum Yaw D gain to tune to.
+    // @User: Standard
+    AP_GROUPINFO("YAW_D_MAX", 7, AP_Quicktune, yaw_d_max, 0.01),
+
+    // @Param: QUIK_RP_PI_RATIO
+    // @DisplayName: Ratio between P and I gains for roll and pitch axes.
+    // @Description:
+    // @User: Standard
+    AP_GROUPINFO("RP_PI_RATIO", 8, AP_Quicktune, rp_pi_ratio, 1.0),
+
+    // @Param: QUIK_Y_PI_RATIO
+    // @DisplayName: Ratio between P and I gains for yaw axis.
+    // @Description:
+    // @User: Standard
+    AP_GROUPINFO("Y_PI_RATIO", 9, AP_Quicktune, y_pi_ratio, 10),
+
+    // @Param: QUIK_AUTO_FILTER
+    // @DisplayName:
+    // @Description:
+    // @User: Standard
+    AP_GROUPINFO("AUTO_FILTER", 10, AP_Quicktune, auto_filter, 1),
+
+    // @Param: QUIK_AUTO_SAVE
+    // @DisplayName:
+    // @Description:
+    // @User: Standard
+    AP_GROUPINFO("AUTO_SAVE", 11, AP_Quicktune, auto_save, 0),
+
+    // @Param: QUIK_MAX_REDUCE
+    // @DisplayName: Maximum reduction in percent.
+    // @Description:
+    // @User: Standard
+    AP_GROUPINFO("MAX_REDUCE", 12, AP_Quicktune, max_reduce, 20),
+
+    // @Param: QUIK_OPTIONS
+    // @DisplayName: Options bitmask.
+    // @Description: Bit 0 uses 2 position switch.
+    // @User: Standard
+    AP_GROUPINFO("OPTIONS", 13, AP_Quicktune, options, 0),
+
+    AP_GROUPEND
+};
+
+// Call at loop rate
+void AP_Quicktune::update(){
+
+    if (enable < 1){
+        return;
+    }
+    if (get_time() - last_update < 1.0/UPDATE_RATE_HZ) {
```

Comment:
need to use AP_HAL::millis() and uint32_t for the intervals for C++ code

Reply:
Done.